internal/provider/provider.go

```diff
@@ -2,70 +2,88 @@ package provider
 
 import (
 	"context"
+	"log"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/kentik/community_sdk_golang/apiv6/kentikapi"
 )
 
+const (
+	apiURLKey      = "apiurl"
+	emailKey       = "email"
+	tokenKey       = "token"
+	logPayloadsKey = "log_payloads"
+)
+
 func init() {
-	// Set descriptions to support markdown syntax, this will be used in document generation and the language server.
+	// Set descriptions to support Markdown syntax, this will be used in document generation and the language server.
 	schema.DescriptionKind = schema.StringMarkdown
 }
 
 func New() *schema.Provider {
-	p := &schema.Provider{
+	return &schema.Provider{
 		Schema: map[string]*schema.Schema{
-			"apiurl": &schema.Schema{
+			apiURLKey: {
 				Type:        schema.TypeString,
-				Optional:    true,
+				Required:    true,
 				DefaultFunc: schema.EnvDefaultFunc("KTAPI_URL", nil),
-				Description: "Custom apiserver url can be specified either by apiurl attribute or KTAPI_URL environment variable. If not specified, default of <https://cloudexports.api.kentik.com> will be used",
+				Description: "CloudExport API server URL (optional). Can also be specified with KTAPI_URL environment variable.",
 			},
-			"email": &schema.Schema{
+			emailKey: {
 				Type:        schema.TypeString,
 				Required:    true,
 				DefaultFunc: schema.EnvDefaultFunc("KTAPI_AUTH_EMAIL", nil),
-				Description: "Authorization. Either email attribute or KTAPI_AUTH_EMAIL environment variable is required",
+				Description: "Authorization email (required). Can also be specified with KTAPI_AUTH_EMAIL environment variable.",
 			},
-			"token": &schema.Schema{
+			tokenKey: {
 				Type:        schema.TypeString,
 				Sensitive:   true,
-				Required:    true,
+				Optional:    true,
```

Comment:
I'm not entirely sure what the `Optional` and `Required` attributes mean here. However if the `emailKey` has `Required` set to true, so should the `tokenKey`. Both email and token must be provided (one way or another) in order for the provider to be able to access the API.

Reply:
Thank you for catching that. I was experimenting with required attributes when I was working on plan_id task in AWS automation and I reverted the changes wrong.